fhir_extractor/src/main.py

```diff
@@ -14,95 +15,141 @@
+            topic = get_topic_name("mimic", resource_type, "extract")
+            producer.produce_event(topic=topic, event=event)
 
-    except Exception as err:
-        logger.error(err)
+    def process_event(msg):
+        msg_value = json.loads(msg.value())
+        resource_id = msg_value.get("resource_id", None)
+        primary_key_values = msg_value.get("primary_key_values", None)
+        batch_id = msg_value.get("batch_id", None)
+
+        msg_topic = msg.topic()
+
+        logger.info("Events Ready to be processed")
+        logger.info(msg_topic)
+        logger.info(msg_value)
+
+        try:
+            resource_mapping, analysis, df = extract_resource(resource_id, primary_key_values)
+            broadcast_events(resource_mapping, df, analysis, batch_id)
+
+        except Exception as err:
+            logger.error(err)
+
+    return process_event
 
 
-def run_resource(resource_mapping, analysis, primary_key_values=None, batch_id=None):
+def manage_kafka_error(msg):
     """
+    Deal with the error if any
+    :param msg:
+    :return:
     """
-    resource_type = resource_mapping["definitionId"]
-    resource_id = resource_mapping["id"]
+    logger.error(msg.error())
+
+
+def extract_resource(resource_id, primary_key_values):
+    logger.debug("Getting Mapping")
+    resource_mapping = get_resource_from_id(resource_id=resource_id)
+
+    # Get credentials
+    if not resource_mapping["source"]["credential"]:
+        raise MissingInformationError("credential is required to run fhir-river by batch.")
 
-    # Extract
+    credentials = resource_mapping["source"]["credential"]
+    extractor.update_connection(credentials)
+
+    logger.debug("Analyzing Mapping")
+    analysis = analyzer.analyze(resource_mapping)
+
+    logger.debug("Extracting rows")
     df = extractor.extract(resource_mapping, analysis, primary_key_values)
     if df.empty:
-        raise ValueError(
+        raise EmptyResult(
             "The sql query returned nothing. Maybe the primary key values "
             "you provided are not present in the database or the mapping "
             "is erroneous."
         )
 
-    list_records_from_db = extractor.split_dataframe(df, analysis)
+    return resource_mapping, analysis, df
 
-    for record in list_records_from_db:
-        logger.debug("One record from extract")
-        event = dict()
-        event["batch_id"] = batch_id
-        event["resource_type"] = resource_type
-        event["resource_id"] = resource_id
-        event["dataframe"] = record.to_dict(orient="list")
 
-        topic = get_topic_name("mimic", resource_type, "extract")
-        producer.produce_event(topic=topic, event=event)
+@app.route("/extract", methods=["POST"])
+def extract():
+    body = request.get_json()
+    resource_id = body.get("resource_id", None)
+    primary_key_values = body.get("primary_key_values", None)
+    if primary_key_values is None or len(primary_key_values) == 0:
```

Comment:
nit: would `if primary_key_values:` be better?

Reply:
what about `if not primary_key_values:` ?

Reply:
Oh, yes, that's what I meant 😅